src/main/ruby/core/posix.rb

```diff
@@ -217,6 +219,47 @@ def self.unsetenv(name)
   else
     raise 'Unsupported platform'
   end
+  
+  THREAD_LOCAL_WRITE_BUFFER = :tuffle_posix_thread_local_write_buffer
+  
+  def self.write_string(fd, string, nonblock=false)
+    length = string.bytesize
+    buffer = Thread.current.thread_variable_get(THREAD_LOCAL_WRITE_BUFFER)
```

Comment:
Using a Thread-variable that is user accessible is nice, but unfortunately I think also dangerous.
I have seen Rails copying Fiber-locals across Fibers, and I could imagine some libraries do the same for Thread-locals across Threads and that would create very hard to debug problems.
Ruby Thread-locals also have quite some overhead, as they need to get the current Thread instance, acquire a lock and they use a Hash for storage.
I think the simplest to avoid these problems would be to have one primitive that just returns an IO buffer (I think we can use it for other operations than `write`, since we can only do one native call at a time) at least of the size given, as a FFI::Pointer or FFI::MemoryPointer or maybe just a raw address which we can wrap in Ruby if that's more convenient.

Reply:
Another option would be to just store the IO buffer on the Ruby Thread instance, and define some private method on Thread to get it with at least the given size.